libraries/rush-lib/src/logic/operations/ShellOperationRunnerPlugin.ts

```diff
@@ -23,54 +24,60 @@ export const PLUGIN_NAME: 'ShellOperationRunnerPlugin' = 'ShellOperationRunnerPl
-      // to specify a runner type requested in rush-project.json
-      const customParameterValues: ReadonlyArray<string> = getCustomParameterValuesForPhase(phase);
-
-      const displayName: string = getDisplayName(phase, project);
-      const { name: phaseName, shellCommand } = phase;
-
-      const { scripts } = project.packageJson;
-
-      // This is the command that will be used to identify the cache entry for this operation
-      const commandForHash: string | undefined = shellCommand ?? scripts?.[phaseName];
-
-      // For execution of non-initial runs, prefer the `:incremental` script if it exists.
-      // However, the `shellCommand` value still takes precedence per the spec for that feature.
-      const commandToRun: string | undefined =
-        shellCommand ??
-        (!isInitial ? scripts?.[`${phaseName}:incremental`] : undefined) ??
-        scripts?.[phaseName];
-
-      operation.runner = initializeShellOperationRunner({
-        phase,
-        project,
-        displayName,
-        commandForHash,
-        commandToRun,
-        customParameterValues,
-        rushConfiguration
-      });
-    }
+    hooks.createOperations.tap(
+      PLUGIN_NAME,
+      function createShellOperations(
+        operations: Set<Operation>,
+        context: ICreateOperationsContext
+      ): Set<Operation> {
+        const { rushConfiguration, isInitial } = context;
+
+        const getCustomParameterValuesForPhase: (phase: IPhase) => ReadonlyArray<string> =
+          getCustomParameterValuesByPhase();
+        for (const operation of operations) {
+          const { associatedPhase: phase, associatedProject: project } = operation;
+
+          if (phase && project && !operation.runner) {
+            // This is a shell command. In the future, may consider having a property on the initial operation
+            // to specify a runner type requested in rush-project.json
+            const customParameterValues: ReadonlyArray<string> = getCustomParameterValuesForPhase(phase);
+
+            const displayName: string = getDisplayName(phase, project);
+            const { name: phaseName, shellCommand } = phase;
+
+            const { scripts } = project.packageJson;
+
+            // This is the command that will be used to identify the cache entry for this operation
+            const commandForHash: string | undefined = shellCommand ?? scripts?.[phaseName];
+
+            // For execution of non-initial runs, prefer the `:incremental` script if it exists.
+            // However, the `shellCommand` value still takes precedence per the spec for that feature.
+            const commandToRun: string | undefined =
+              shellCommand ??
+              (!isInitial ? scripts?.[`${phaseName}:incremental`] : undefined) ??
+              scripts?.[phaseName];
+
+            const environment: IEnvironment | undefined = hooks.createEnvironmentForOperation.isUsed()
```

Comment:
This would be more generally useful if it was invoked _during_ invocation of the runner, instead of in advance. An example use case is provided in #5094 , which needs a way to tell a runner about assigned resources.

Reply:
Fixed so that it late binds.